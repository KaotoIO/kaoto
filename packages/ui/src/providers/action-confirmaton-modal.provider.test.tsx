import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import { FunctionComponent, useContext } from 'react';
import {
  ACTION_ID_CANCEL,
  ACTION_ID_CONFIRM,
  ActionConfirmationButtonOption,
  ActionConfirmationModalContext,
  ActionConfirmationModalContextProvider,
} from './action-confirmation-modal.provider';
import { ButtonVariant } from '@patternfly/react-core';

let actionConfirmationResult: string | undefined;

describe('ActionConfirmationModalProvider', () => {
  beforeEach(() => {
    actionConfirmationResult = undefined;
  });

  it('calls actionConfirmation with true when Confirm button is clicked', async () => {
    render(
      <ActionConfirmationModalContextProvider>
        <TestComponent title="Permanently delete step" text="Step parameters and its children will be lost." />
      </ActionConfirmationModalContextProvider>,
    );

    const deleteButton = screen.getByText('Delete');
    fireEvent.click(deleteButton);

    const confirmButton = await screen.findByRole('button', { name: 'Confirm' });
    fireEvent.click(confirmButton);

    // Wait for actionConfirmation promise to resolve
    await waitFor(() => expect(actionConfirmationResult).toEqual(ACTION_ID_CONFIRM));
  });

  it('calls actionConfirmation with false when Cancel button is clicked', async () => {
    render(
      <ActionConfirmationModalContextProvider>
        <TestComponent title="Permanently delete step" text="Step parameters and its children will be lost." />
      </ActionConfirmationModalContextProvider>,
    );

    const deleteButton = screen.getByText('Delete');
    fireEvent.click(deleteButton);

    const cancelButton = screen.getByRole('button', { name: 'Cancel' });
    fireEvent.click(cancelButton);

    // Wait for actionConfirmation promise to resolve
    await waitFor(() => expect(actionConfirmationResult).toEqual(ACTION_ID_CANCEL));
  });

  it('should allow consumers to update the modal title and text', () => {
    const wrapper = render(
      <ActionConfirmationModalContextProvider>
        <TestComponent title="Custom title" text="Custom text" />
      </ActionConfirmationModalContextProvider>,
    );

    act(() => {
      const deleteButton = wrapper.getByText('Delete');
      fireEvent.click(deleteButton);
    });

    const modalDialog = wrapper.getByRole('dialog');
    expect(modalDialog).toMatchSnapshot();

    expect(wrapper.queryByText('Custom title')).toBeInTheDocument;
    expect(wrapper.queryByText('Custom text')).toBeInTheDocument;
  });

  it('should show 3 options to choose', async () => {
    const wrapper = render(
      <ActionConfirmationModalContextProvider>
        <TestComponent
          title="Custom title"
          text="Custom text"
          additionalModalText="Additional text is added in the modal description"
          buttonOptions={{
            'del-step-and-file': {
              buttonText: 'Delete the step, and delete the file(s)',
              variant: ButtonVariant.danger,
            },
            'del-step-only': {
              buttonText: 'Delete the step, but keep the file(s)',
              variant: ButtonVariant.secondary,
              isDanger: true,
            },
          }}
        />
      </ActionConfirmationModalContextProvider>,
    );

    act(() => {
      const deleteButton = wrapper.getByText('Delete');
      fireEvent.click(deleteButton);
    });
    const modalDialog = wrapper.getByRole('dialog');
    expect(modalDialog.textContent).toContain('Additional text is added in the modal description');
    act(() => {
      const cancelButton = wrapper.getByTestId('action-confirmation-modal-btn-cancel');
      expect(cancelButton.textContent).toEqual('Cancel');
      fireEvent.click(cancelButton);
    });
    await waitFor(() => {
      expect(actionConfirmationResult).toEqual(ACTION_ID_CANCEL);
    });

    act(() => {
      const deleteButton = wrapper.getByText('Delete');
      fireEvent.click(deleteButton);
    });
    act(() => {
      const deleteStepAndFileButton = wrapper.getByTestId('action-confirmation-modal-btn-del-step-and-file');
      expect(deleteStepAndFileButton.textContent).toEqual('Delete the step, and delete the file(s)');
      fireEvent.click(deleteStepAndFileButton);
    });
    await waitFor(() => {
      expect(actionConfirmationResult).toEqual('del-step-and-file');
    });

    act(() => {
      const deleteButton = wrapper.getByText('Delete');
      fireEvent.click(deleteButton);
    });
    act(() => {
      const deleteStepOnlyButton = wrapper.getByTestId('action-confirmation-modal-btn-del-step-only');
      expect(deleteStepOnlyButton.textContent).toEqual('Delete the step, but keep the file(s)');
      fireEvent.click(deleteStepOnlyButton);
    });
    await waitFor(() => {
      expect(actionConfirmationResult).toEqual('del-step-only');
    });
  });
});

interface TestComponentProps {
  title: string;
  text: string;
  additionalModalText?: string;
  buttonOptions?: Record<string, ActionConfirmationButtonOption>;
}

const TestComponent: FunctionComponent<TestComponentProps> = (props) => {
  const { actionConfirmation: deleteConfirmation } = useContext(ActionConfirmationModalContext)!;

  const handleDelete = async () => {
    const confirmation = await deleteConfirmation(props);
    actionConfirmationResult = confirmation;
  };

  return <button onClick={handleDelete}>Delete</button>;
};
