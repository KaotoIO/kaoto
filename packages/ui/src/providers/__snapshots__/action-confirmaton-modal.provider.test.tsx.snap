// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionConfirmationModalProvider should allow consumers to update the modal title and text 1`] = `
<div
  aria-describedby="pf-modal-part-4"
  aria-labelledby="pf-modal-part-3"
  aria-modal="true"
  class="pf-v5-c-modal-box pf-m-warning pf-m-sm"
  data-ouia-component-id="ActionConfirmationModal"
  data-ouia-component-type="PF5/ModalContent"
  data-ouia-safe="true"
  id="pf-modal-part-2"
  role="dialog"
>
  <div
    class="pf-v5-c-modal-box__close"
  >
    <button
      aria-disabled="false"
      aria-label="Close"
      class="pf-v5-c-button pf-m-plain"
      data-ouia-component-id="ActionConfirmationModal-ModalBoxCloseButton"
      data-ouia-component-type="PF5/Button"
      data-ouia-safe="true"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="pf-v5-svg"
        fill="currentColor"
        height="1em"
        role="img"
        viewBox="0 0 352 512"
        width="1em"
      >
        <path
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        />
      </svg>
    </button>
  </div>
  <header
    class="pf-v5-c-modal-box__header"
  >
    <h1
      class="pf-v5-c-modal-box__title pf-m-icon"
      id="pf-modal-part-3"
    >
      <span
        class="pf-v5-c-modal-box__title-icon"
      >
        <svg
          aria-hidden="true"
          class="pf-v5-svg"
          fill="currentColor"
          height="1em"
          role="img"
          viewBox="0 0 576 512"
          width="1em"
        >
          <path
            d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          />
        </svg>
      </span>
      <span
        class="pf-v5-screen-reader"
      >
        Warning alert:
      </span>
      <span
        class="pf-v5-c-modal-box__title-text"
      >
        Custom title
      </span>
    </h1>
  </header>
  <div
    class="pf-v5-c-modal-box__body"
    id="pf-modal-part-4"
  >
    Custom text
  </div>
  <footer
    class="pf-v5-c-modal-box__footer"
  >
    <div
      class="pf-v5-l-split pf-m-gutter pf-m-wrap"
    >
      <div
        class="pf-v5-l-split__item"
      >
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-danger"
          data-ouia-component-id="OUIA-Generated-Button-danger-3"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          data-testid="action-confirmation-modal-btn-confirm"
          type="button"
        >
          Confirm
        </button>
      </div>
      <div
        class="pf-v5-l-split__item"
      >
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-link"
          data-ouia-component-id="OUIA-Generated-Button-link-3"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          data-testid="action-confirmation-modal-btn-cancel"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </footer>
</div>
`;
